svcb.go

```diff
@@ -334,13 +334,56 @@ func (s *SVCBMandatory) copy() SVCBKeyValue {
 //	h.Hdr = dns.RR_Header{Name: ".", Rrtype: dns.TypeHTTPS, Class: dns.ClassINET}
 //	e := new(dns.SVCBAlpn)
 //	e.Alpn = []string{"h2", "http/1.1"}
-//	h.Value = append(o.Value, e)
+//	h.Value = append(h.Value, e)
 type SVCBAlpn struct {
 	Alpn []string
 }
 
-func (*SVCBAlpn) Key() SVCBKey     { return SVCB_ALPN }
-func (s *SVCBAlpn) String() string { return strings.Join(s.Alpn, ",") }
+func (*SVCBAlpn) Key() SVCBKey { return SVCB_ALPN }
+
+func (s *SVCBAlpn) String() string {
+	// An ALPN value is a comma-separated list of values, each of which can be
+	// an arbitrary binary value. In order to allow parsing, the comma and
+	// backslash characters are themselves excaped.
+	//
+	// However, this escaping is done in addition to the normal escaping which
+	// happens in zone files, meaning that these values must be
+	// double-escaped. This looks terrible, so if you see a never-ending
+	// sequence of backslash in a zone file this may be why.
+	//
+	// https://datatracker.ietf.org/doc/html/draft-ietf-dnsop-svcb-https-08#appendix-A.1
+	esc := make([]string, len(s.Alpn))
+	for i, alpn := range s.Alpn {
+		var str strings.Builder
+		str.Grow(4 * len(alpn))
+		for j := 0; j < len(alpn); j++ {
+			e := alpn[j]
+			if ' ' <= e && e <= '~' {
+				switch e {
+				// We escape a few characters which may confuse humans or
+				// parsers.
+				case '"', ';', ' ':
+					str.WriteByte('\\')
+					str.WriteByte(e)
+				// The comma and backslash characters themselves must be
+				// doubly-escaped. We use `\\` for the first backslash and
+				// the escaped numeric value for the other value. We especially
+				// don't want a comma in the output.
+				case ',':
+					str.WriteString(`\\\044`)
+				case '\\':
+					str.WriteString(`\\\092`)
+				default:
+					str.WriteByte(e)
+				}
+			} else {
```

Comment:
reverse this `if`, so the else part is just done +  a `continue` and the current if-body can be pulled to the left

Reply:
Makes sense, done.